src/instantiate.jl

```diff
@@ -0,0 +1,116 @@
+"""
+    struct OptimizerWithAttributes
+        optimizer
+        params::Vector{Pair{AbstractOptimizerAttribute,<:Any}}
+    end
+
+Object grouping an optimizer constructor and a list of optimizer attributes.
+Instances are created with [`instantiate`](@ref).
+"""
+struct OptimizerWithAttributes
+    # Function that takes zero arguments and returns a new optimizer.
+    # The type of the function could be
+    # * `Function`: a function, or
+    # * `DataType`: a type, or
+    # * `UnionAll`: a type with missing parameters.
+    optimizer_constructor
+    params::Vector{Pair{AbstractOptimizerAttribute,Any}}
+end
+
+_to_param(param::Pair{<:AbstractOptimizerAttribute}) = param
+_to_param(param::Pair{String}) = RawParameter(param.first) => param.second
+function _to_param(param::Pair)
+    error("Expected an optimizer attribute or a string, got `$(param.first)` which is a `$(typeof(param.first))`.")
+end
+
+"""
+    OptimizerWithAttributes(optimizer_constructor, params::Pair...)
+
+Create an [`OptimizerWithAttributes`](@ref) with the parameters `params`.
+"""
+function OptimizerWithAttributes(optimizer_constructor, args::Vararg{Pair, N}) where N
+    if !applicable(optimizer_constructor)
+        error(_INSTANTIATE_NOT_CALLABLE_MESSAGE)
+    end
+    params = Pair{AbstractOptimizerAttribute,Any}[_to_param(arg) for arg in args]
+    return OptimizerWithAttributes(optimizer_constructor, params)
+end
+
+const _INSTANTIATE_NOT_CALLABLE_MESSAGE =
+    "The provided `optimizer_constructor` is invalid. It must be callable with zero" *
+    "arguments. For example, \"Ipopt.Optimizer\" or" *
+    "\"() -> ECOS.Optimizer()\". It should not be an instantiated optimizer " *
+    "like \"Ipopt.Optimizer()\" or \"ECOS.Optimizer()\"." *
+    "(Note the difference in parentheses!)"
```

Comment:
Needs spaces at end of each line otherwise the text runs together:
```Julia
julia> _INSTANTIATE_NOT_CALLABLE_MESSAGE
"The provided `optimizer_constructor` is invalid. It must be callable with zeroarguments. For example, \"Ipopt.Optimizer\" or\"() -> ECOS.Optimizer()\". It should not be an instantiated optimizer like \"Ipopt.Optimizer()\" or \"ECOS.Optimizer()\".(Note the difference in parentheses!)"
```